### PR TITLE
3 stream to ipa

### PR DESCRIPTION
**Pull request by arunasrivastava**
PR for decoupled streaming logic

### CODE REVIEW

File: src/transcription.py [hunk 1]
```diff
@@ -1,53 +1,75 @@
...
 
 # Load Wav2Vec2 model
 model_id = "KoelLabs/xlsr-english-01"
-processor = AutoProcessor.from_pretrained(model_id)
-model = AutoModelForCTC.from_pretrained(model_id)
+processor: Wav2Vec2Processor = AutoProcessor.from_pretrained(model_id)
+model: Wav2Vec2ForCTC = AutoModelForCTC.from_pretrained(model_id)
 assert processor.feature_extractor.sampling_rate == SAMPLE_RATE
+MIN_LEN_SAMPLES = (
+    400  # computed from model.config.conv_kernel and model.config.conv_stride
```
Comment (SanderGi) [inline src/transcription.py:19]: Nit: as mentioned above (with suggested code), you probably want to compute it directly rather than explain in a comment how it could be computed

File: src/transcription.py [hunk 2]
```diff
@@ -1,53 +1,75 @@
...
+        conv_feats_t = conv_feats.transpose(1, 2)  # (B, T', C)
+        # Project to hidden size for transformer; also returns normalized conv features
+        features, normed_conv_feats = model.wav2vec2.feature_projection(conv_feats_t)
+    # Return transformer-ready features and original (unpadded) input length in samples
+    return features, raw_sample_count
 
-    predicted_ids = torch.argmax(logits, dim=-1)[0].tolist()
-    duration_sec = input_values.shape[1] / processor.feature_extractor.sampling_rate
 
+def run_transformer_on_features(features, total_audio_samples, time_offset=0.0):
```
Comment (SanderGi) [inline src/transcription.py:45]: Nit: put back the return type ` -> TIMESTAMPED_PHONES_T`

File: src/transcription.py [hunk 3]
```diff
@@ -1,53 +1,75 @@
...
-            padding=True,
-        )
-        .input_values.type(torch.float32)
-        .to(model.device)
+def extract_features_only(audio: np.ndarray):
+    """Extract CNN features and project to encoder hidden size (transformer-ready)."""
+    # True raw sample count before any padding
+    raw_sample_count = int(np.asarray(audio).shape[-1])
+    if raw_sample_count < MIN_LEN_SAMPLES:
+        audio = np.pad(audio, (0, MIN_LEN_SAMPLES - raw_sample_count), mode="constant")
```
Comment (SanderGi) [inline src/transcription.py:28]: You probably want to use the processor to pad these so they also get marked as padding in the attention mask

File: src/transcription.py [hunk 4]
```diff
@@ -1,53 +1,75 @@
...
+    return features, raw_sample_count
 
-    predicted_ids = torch.argmax(logits, dim=-1)[0].tolist()
-    duration_sec = input_values.shape[1] / processor.feature_extractor.sampling_rate
 
+def run_transformer_on_features(features, total_audio_samples, time_offset=0.0):
+    """Run transformer and decode"""
+    #  slowest step
+    with torch.no_grad():
+        encoder_outputs = model.wav2vec2.encoder(features)
```
Comment (SanderGi) [inline src/transcription.py:49]: You are getting rid of the attention mask which I believe will cause the padding tokens to be attended to

File: src/server.py [hunk 5]
```diff
@@ -83,44 +89,48 @@ def get_score_words_cer():
...
-
-                    wavfile.write("src/audio.wav", SAMPLE_RATE, audio)
-                    combined = np.concatenate([combined, audio])
-                    wavfile.write("src/combined.wav", SAMPLE_RATE, combined)
-
-                if data == "stop":
-                    break
-
-                buffer = b""  # Clear the buffer
+            # Process 20ms chunks
```
Comment (SanderGi) [inline src/server.py:102]: The 20ms is the stride. The CNN actually has a receptive field of 25ms (400 samples). By only ever feeding it 20ms chunks and padding them to fit the 25ms (`MIN_LEN_SAMPLES`), you never allow it to read all the audio it needs to perform equivalently to running the full model once at the end. You'll want to let it collect some multiple of 320 samples + 80 samples in each chunk or, if you prefer to allow the last CNN receptive field to be incomplete, you should recompute it when more audio comes in

File: src/server.py [hunk 6]
```diff
@@ -12,12 +12,18 @@
...
+    SAMPLE_RATE,
+    transcribe_timestamped,
+)
 from phoneme_utils import TIMESTAMPED_PHONES_T, TIMESTAMPED_PHONES_BY_WORD_T
 
 # Constants
 DEBUG = False
-NUM_SECONDS_PER_CHUNK = 0.5
+CHUNK_SIZE_SAMPLES = 320  # 20ms at 16kHz
+TRANSFORMER_INTERVAL = 30
```
Comment (SanderGi) [inline src/server.py:26]: We'll definitely want this to be adaptive. Not an immediate priority. Make sure you write good benchmarking code first so you can measure whether each change is an improvement. Then we can also combine with the VAD and local agreement optimizations

File: src/server.py [hunk 7]
```diff
@@ -12,12 +12,18 @@
     top_phonetic_errors,
     pair_by_words,
 )
-from transcription import transcribe_timestamped, SAMPLE_RATE
+from transcription import (
+    extract_features_only,
+    run_transformer_on_features,
+    SAMPLE_RATE,
+    transcribe_timestamped,
```
Comment (SanderGi) [inline src/server.py:19]: Nit: you removed `transcribe_timestamped` so it can no longer be imported. Also `SAMPLE_RATE` is an unused import

File: src/server.py [hunk 8]
```diff
@@ -12,12 +12,18 @@
...
+)
 from phoneme_utils import TIMESTAMPED_PHONES_T, TIMESTAMPED_PHONES_BY_WORD_T
 
 # Constants
 DEBUG = False
-NUM_SECONDS_PER_CHUNK = 0.5
+CHUNK_SIZE_SAMPLES = 320  # 20ms at 16kHz
```
[the remark this answers is not in the document]
Comment (arunasrivastava) [reply src/server.py:25]: is there zero performance benefit from hardcoding

[the remark this answers is not in the document]
Comment (SanderGi) [reply src/server.py]: Not in any way that matters. If you just compute these constants once on startup and store them into the same module level variables, you are looking at 5.7% of a millisecond of extra server startup time. A handful of multiplications like this are so fast that you wouldn't even be able to perceive it if you re-ran this computation on every chunk (no need to do this though). Most compiled languages will be able to do some optimizations on compile-time constants (such as constant folding, division through reciprocal multiplication, etc.). However this is not something Python really does due to its more interpreted nature. Moreover, even for compiled languages, it is almost never the right move to sacrifice readability/flexibility for the minuscule gain of hardcoding everything. I encourage you to run some experiments and measure the performance yourself

File: src/server.py (hunk 6 above)
Comment (SanderGi) [reply src/server.py:26]: Wrote some suggested utils for benchmarking code [here](https://github.com/KoelLabs/ML/blob/main/scripts/ipa_transcription/wav2vec2_streaming.py). Make sure to add accuracy, average over multiple samples, and create nice figures for a blog post
